core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java

```diff
@@ -0,0 +1,172 @@
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator.Balancer;
+import org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator.RebalanceDecision;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDecider;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDeciders;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision.Type;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Tests for balancing a single shard, see {@link Balancer#balanceShard(ShardRouting)}.
+ */
+public class BalancedSingleShardTests extends ESAllocationTestCase {
+
+    public void testRebalanceNonStartedShardNotAllowed() {
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomBoolean(),
+            randomFrom(ShardRoutingState.INITIALIZING, ShardRoutingState.UNASSIGNED));
+        ShardRouting unassignedShard = clusterState.routingTable().index("idx").shard(0).primaryShard();
+        RebalanceDecision rebalanceDecision = allocator.rebalanceShard(unassignedShard, newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.emptyList()), clusterState));
+        assertSame(RebalanceDecision.NOT_TAKEN, rebalanceDecision);
+    }
+
+    public void testRebalanceNotAllowedDuringPendingAsyncFetch() {
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomBoolean(), ShardRoutingState.STARTED);
+        ShardRouting unassignedShard = clusterState.routingTable().index("idx").shard(0).primaryShard();
+        RoutingAllocation routingAllocation = newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.emptyList()), clusterState);
+        routingAllocation.setHasPendingAsyncFetch();
+        RebalanceDecision rebalanceDecision = allocator.rebalanceShard(unassignedShard, routingAllocation);
+        assertNotNull(rebalanceDecision.getCanRebalanceDecision());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertEquals("cannot rebalance due to in-flight shard store fetches", rebalanceDecision.getFinalExplanation());
+        assertNull(rebalanceDecision.getNodeDecisions());
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testRebalanceNotAllowed() {
+        AllocationDecider noRebalanceDecider = new AllocationDecider(Settings.EMPTY) {
```

Comment:
`noAllocationDeciders` and `throttleAllocationDeciders` can be used instead here